README.md
# Open Spacedeck Plus

This package is forked from open space deck to make it usuable in productive use.

# Features

- Create virtual whiteboards called *Spaces* with virtually unlimited size
- Drag & drop images, videos and audio from your computer or the web
- Write and format text with full control over fonts, colors and style
- Draw, annotate and highlight with included graphical shapes
- Turn your Space into a zooming presentation
- Collaborate and chat in realtime with teammates, students or friends
- Share Spaces on the web or via email
- Export your work as printable PDF or ZIP

# Use Cases

- Education: Virtual classwork with multimedia
- Creative: Mood boards, Brainstorming, Design Thinking
- Visual note taking and planning



# Data Import from Spacedeck.com

Spacedeck Open has a data import feature that you can use to migrate your ZIP export from Spacedeck.com.

1. Just copy your downloaded ZIP file into the spacedeck root folder. Don't extract it.
2. Start your local Spacedeck.
3. Navigate to *Account / Profile* (person icon in the top right corner).
4. Click the *Import* button next to the ZIP file name. It is on the bottom of the page.
5. Wait until console output has finished and you're done.

# Requirements, Installation

Spacedeck requires:

- Node.js 9.x: Web Server / API. Download: https://nodejs.org

To run Spacedeck, you only need Node.JS 9.x.

To install all node dependencies, run (do this once):

    npm install

# Configuration

See [config/default.json](config/default.json)

# Run (web server)

    node spacedeck.js

Then open http://localhost:9666 in a web browser.

# Run (desktop app with integrated web server)

    electron .

# Optional Dependencies

For advanced media conversion:

- ffmpeg and ffprobe for video/audio conversion. Download: https://www.ffmpeg.org/download.html
- audiowaveform for audio waveform rendering. Download: https://github.com/bbcrd/audiowaveform
- ghostscript for PDF import. Download: https://www.ghostscript.com/download/gsdnld.html

# Data Storage

By default, media files are uploaded to the ```storage``` folder.
The database is stored in ```database.sqlite``` by default.

# Hacking

To rebuild the frontend CSS styles:

    gulp styles

# License

The Spacedeck logo and brand assets are registered trademarks of Spacedeck GmbH. All rights reserved.

Spacedeck Open source code is released under the GNU Affero General Public License Version 3 (GNU AGPLv3).

    Spacedeck Open - Web-based Collaborative Whiteboard For Rich Media
    Copyright (C) 2011-2018 Lukas F. Hartmann, Martin Güther
    Icons and original CSS design copyright by Thomas Helbig
    
    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU Affero General Public License as
    published by the Free Software Foundation, either version 3 of the
    License, or (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Affero General Public License for more details.

    You should have received a copy of the GNU Affero General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.


changelog.md

# 6/11/2018

Added autocreate endpoint for public Whiteboards only identified by a long hash;

e.g. 

    http://localhost:9666/autocreate/8a080f08-e2ea-4ffc-abc9-0328cf675142/  
    
Will create or open the board;


models/db.js
const Umzug = require('umzug');
const _ = require('lodash');

function sequel_log(a, b, c) {
    console.log(a);
}

const Sequelize = require('sequelize');
const config = require('config');

const sequlizeDefaults = {
    host: 'localhost',
    dialect: 'sqlite',

    pool: {
        max: 5,
        min: 0,
        acquire: 30000,
        idle: 10000
    },

    // SQLite only
    storage: 'database.sqlite',
    logging: sequel_log,

    // http://docs.sequelizejs.com/manual/tutorial/querying.html#operators
    operatorsAliases: false
};

const sequlizeOptions =  _.extend(sequlizeDefaults, config.db.sequlizeOptions);
const sequelize = new Sequelize( config.db.database , config.db.username, config.db.password, sequlizeOptions);

var User;
var Session;
var Space;
var Membership;
var Artifact;
var Message;
var Action;

module.exports = {
    User: sequelize.define('user', {
        _id: { type: Sequelize.STRING, primaryKey: true },
        email: Sequelize.STRING,
        password_hash: Sequelize.STRING,
        nickname: Sequelize.STRING,
        avatar_original_uri: Sequelize.STRING,
        avatar_thumb_uri: Sequelize.STRING,
        confirmation_token: Sequelize.STRING,
        password_reset_token: Sequelize.STRING,
        home_folder_id: Sequelize.STRING,
        prefs_language: Sequelize.STRING,
        prefs_email_notifications: Sequelize.STRING,
        prefs_email_digest: Sequelize.STRING,
        created_at: { type: Sequelize.DATE, defaultValue: Sequelize.NOW },
        updated_at: { type: Sequelize.DATE, defaultValue: Sequelize.NOW }
    }),

    Session: sequelize.define('session', {
        token: { type: Sequelize.STRING, primaryKey: true },
        user_id: Sequelize.STRING,
        expires: Sequelize.DATE,
        created_at: { type: Sequelize.DATE, defaultValue: Sequelize.NOW },
        device: Sequelize.STRING,
        ip: Sequelize.STRING
    }),

    Space: sequelize.define('space', {
        _id: { type: Sequelize.STRING, primaryKey: true },
        name: { type: Sequelize.STRING, default: "New Space" },
        space_type: { type: Sequelize.STRING, defaultValue: "space" },
        creator_id: Sequelize.STRING,
        parent_space_id: Sequelize.STRING,

        access_mode: { type: Sequelize.STRING, default: "private" }, // "public" || "private"
        password: Sequelize.STRING,
        edit_hash: Sequelize.STRING,
        edit_slug: Sequelize.STRING,
        editors_locking: Sequelize.BOOLEAN,

        thumbnail_uri: Sequelize.STRING,

        width: Sequelize.INTEGER,
        height: Sequelize.INTEGER,
        background_color: Sequelize.STRING,
        background_uri: Sequelize.STRING,

        created_at: { type: Sequelize.DATE, defaultValue: Sequelize.NOW },
        updated_at: { type: Sequelize.DATE, defaultValue: Sequelize.NOW },
        thumbnail_url: Sequelize.STRING,
        thumbnail_updated_at: { type: Sequelize.DATE }
    }),

    Membership: sequelize.define('membership', {
        _id: { type: Sequelize.STRING, primaryKey: true },
        space_id: Sequelize.STRING,
        user_id: Sequelize.STRING,
        role: Sequelize.STRING,
        code: Sequelize.STRING,
        state: { type: Sequelize.STRING, defaultValue: "pending" },
        created_at: { type: Sequelize.DATE, defaultValue: Sequelize.NOW },
        updated_at: { type: Sequelize.DATE, defaultValue: Sequelize.NOW }
    }),

    Message: sequelize.define('message', {
        _id: { type: Sequelize.STRING, primaryKey: true },
        space_id: Sequelize.STRING,
        user_id: Sequelize.STRING,
        editor_name: Sequelize.STRING,
        message: Sequelize.TEXT,
        created_at: { type: Sequelize.DATE, defaultValue: Sequelize.NOW },
        updated_at: { type: Sequelize.DATE, defaultValue: Sequelize.NOW }
    }),

    Artifact: sequelize.define('artifact', {
        _id: { type: Sequelize.STRING, primaryKey: true },
        space_id: Sequelize.STRING,
        user_id: Sequelize.STRING,

        mime: Sequelize.STRING,
        thumbnail_uri: Sequelize.STRING,
        last_update_user_id: Sequelize.STRING,
        editor_name: Sequelize.STRING,
        last_update_editor_name: Sequelize.STRING,
        description: Sequelize.TEXT,
        state: { type: Sequelize.STRING, default: "idle" },

        //linked_to: Sequelize.STRING,
        title: Sequelize.STRING,
        tags: Sequelize.TEXT,
        search_text: Sequelize.STRING,
        link_uri: Sequelize.STRING,
        play_from: Sequelize.DECIMAL,
        play_to: Sequelize.DECIMAL,

        x: { type: Sequelize.DECIMAL, default: 0.0 },
        y: { type: Sequelize.DECIMAL, default: 0.0 },
        z: { type: Sequelize.DECIMAL, default: 0.0 },
        r: { type: Sequelize.DECIMAL, default: 0.0 },
        w: { type: Sequelize.DECIMAL, default: 100 },
        h: { type: Sequelize.DECIMAL, default: 100 },

        //control_points: [{
        //  dx: Number, dy: Number
        //}],

        control_points: Sequelize.TEXT,

        group: Sequelize.STRING,
        locked: { type: Sequelize.BOOLEAN, default: false },

        payload_uri: Sequelize.STRING,
        payload_thumbnail_web_uri: Sequelize.STRING,
        payload_thumbnail_medium_uri: Sequelize.STRING,
        payload_thumbnail_big_uri: Sequelize.STRING,
        payload_size: Sequelize.INTEGER, // file size in bytes

        fill_color: { type: Sequelize.STRING, default: "transparent" },
        stroke_color: { type: Sequelize.STRING, default: "#000000" },
        text_color: Sequelize.STRING,
        stroke: { type: Sequelize.DECIMAL, default: 0.0 },
        stroke_style: { type: Sequelize.STRING, default: "solid" },
        alpha: { type: Sequelize.DECIMAL, default: 1.0 },
        order: { type: Sequelize.INTEGER, default: 0 },
        crop_x: Sequelize.INTEGER,
        crop_y: Sequelize.INTEGER,
        crop_w: Sequelize.INTEGER,
        crop_h: Sequelize.INTEGER,
        shape: Sequelize.STRING,
        shape_svg: Sequelize.TEXT,
        padding_left: Sequelize.INTEGER,
        padding_right: Sequelize.INTEGER,
        padding_top: Sequelize.INTEGER,
        padding_bottom: Sequelize.INTEGER,
        margin_left: Sequelize.INTEGER,
        margin_right: Sequelize.INTEGER,
        margin_top: Sequelize.INTEGER,
        margin_bottom: Sequelize.INTEGER,
        border_radius: Sequelize.INTEGER,
        align: { type: Sequelize.STRING, default: "left" },
        valign: { type: Sequelize.STRING, default: "top" },

        brightness: Sequelize.DECIMAL,
        contrast: Sequelize.DECIMAL,
        saturation: Sequelize.DECIMAL,
        blur: Sequelize.DECIMAL,
        hue: Sequelize.DECIMAL,
        opacity: Sequelize.DECIMAL,

        payload_alternatives: Sequelize.TEXT,

        /*payload_alternatives: [{
          mime: String,
          payload_uri: String,
          payload_thumbnail_web_uri: String,
          payload_thumbnail_medium_uri:  String,
          payload_thumbnail_big_uri: String,
          payload_size: Number
        }],*/

        created_at: { type: Sequelize.DATE, defaultValue: Sequelize.NOW },
        updated_at: { type: Sequelize.DATE, defaultValue: Sequelize.NOW }
    }),

    init: async function () {
        User = this.User;
        Session = this.Session;
        Space = this.Space;
        Artifact = this.Artifact;
        Message = this.Message;
        Membership = this.Membership;

        Space.belongsTo(User, {
            foreignKey: {
                name: 'creator_id'
            },
            as: 'creator'
        });

        Membership.belongsTo(User, {
            foreignKey: {
                name: 'user_id'
            },
            as: 'user'
        });

        Membership.belongsTo(Space, {
            foreignKey: {
                name: 'space_id'
            },
            as: 'spac'
        });

        Artifact.belongsTo(User, {
            foreignKey: {
                name: 'user_id'
            },
            as: 'user'
        });

        Artifact.belongsTo(Space, {
            foreignKey: {
                name: 'space_id'
            },
            as: 'space'
        });

        Message.belongsTo(User, {
            foreignKey: {
                name: 'user_id'
            },
            as: 'user'
        });

        Message.belongsTo(Space, {
            foreignKey: {
                name: 'space_id'
            },
            as: 'space'
        });

        await sequelize.sync();

        var umzug = new Umzug({
            storage: 'sequelize',
            storageOptions: {
                sequelize: sequelize
            },
            migrations: {
                params: [
                    sequelize.getQueryInterface(),
                    Sequelize
                ],
                path: './models/migrations',
                pattern: /\.js$/
            }
        });

        umzug.up().then(function (migrations) {
            console.log('Migration complete up!');
        });

    },

    getUserRoleInSpacePromise: (originalSpace, user) => {
        const self = this;
        return new Promise((resolve, reject) => {
            self.getUserRoleInSpace(originalSpace, user, (res) => resolve(res));
        });
    },
    getUserRoleInSpace: (originalSpace, user, cb) => {
        originalSpace.path = [];
        console.log("getUserRoleInSpace", originalSpace._id, user._id, user.home_folder_id);

        if (originalSpace._id == user.home_folder_id || (originalSpace.creator_id && originalSpace.creator_id == user._id)) {
            cb("admin");
            return;
        }
        var findMembershipsForSpace = function (space, allMemberships, prevRole) {
            Membership.findAll({
                where: {
                    "space_id": space._id
                }
            }).then(function (parentMemberships) {
                var currentMemberships = parentMemberships.concat(allMemberships);

                if (space.parent_space_id) {
                    Space.findOne({
                        where: {
                            "_id": space.parent_space_id
                        }
                    }, function (err, parentSpace) {
                        findMembershipsForSpace(parentSpace, currentMemberships, prevRole);
                    });
                } else {
                    // reached the top
                    var role = prevRole;
                    space.memberships = currentMemberships;

                    if (role == "none") {
                        if (originalSpace.access_mode == "public") {
                            role = "viewer";
                        }
                    }

                    currentMemberships.forEach(function (m, i) {
                        if (m.user_id && m.user_id == user._id) {
                            role = m.role;
                        }
                    });

                    cb(role);
                }
            });
        };
        findMembershipsForSpace(originalSpace, [], "none");

    },

    spaceToObject: (space) => {
        // FIXME TODO
        return space;
    },

    findUserBySessionToken: (token, cb) => {
        Session.findOne({ where: { token: token } })
            .then(session => {
                if (!session) cb(null, null)
                else User.findOne({ where: { _id: session.user_id } })
                    .then(user => {
                        cb(null, user)
                    })
            })
    },

    unpackArtifact: (a) => {
        if (a.tags && (typeof a.tags) == "string") {
            a.tags = JSON.parse(a.tags);
        }
        if (a.control_points && (typeof a.control_points) == "string") {
            a.control_points = JSON.parse(a.control_points);
        }
        if (a.payload_alternatives && (typeof a.payload_alternatives) == "string") {
            a.payload_alternatives = JSON.parse(a.payload_alternatives);
        }
        return a;
    },

    packArtifact: (a) => {
        if (a.tags && (typeof a.tags) != "string") {
            a.tags = JSON.stringify(a.tags);
        }
        if (a.control_points && (typeof a.control_points) != "string") {
            a.control_points = JSON.stringify(a.control_points);
        }
        if (a.payload_alternatives && (typeof a.payload_alternatives) != "string") {
            a.payload_alternatives = JSON.stringify(a.payload_alternatives);
        }
        return a;
    }
}


webpack.config.js
const MergeIntoSingleFilePlugin = require('webpack-merge-and-include-globally');
const CopyWebpackPlugin = require('copy-webpack-plugin')


module.exports = {
    entry: {
        main: './public/javascripts/spacedeck_whiteboard.js'
    },
    plugins: [
        new MergeIntoSingleFilePlugin({
            files: {
                "javascripts/spacedeck.js": [
                    'public/javascripts/jquery-2.1.4.min.js',
                    'public/javascripts/i18next-1.11.2.js',
                    'public/javascripts/clipboard.js',

                    'public/javascripts/lodash.compat.js',
                    'public/javascripts/fastclick.js',
                    'public/javascripts/vue.js',
                    'public/javascripts/moment.js',
                    'public/javascripts/medium.patched.js',
                    'public/javascripts/route-recognizer.js',

                    'public/javascripts/backend.js',
                    'public/javascripts/link_parser.js',
                    'public/javascripts/vector-render.js',
                    'public/javascripts/mousetrap.js',
                    'public/javascripts/smoke.js',
                    'public/javascripts/helper.js',
                    'public/javascripts/packer.growing.js',

                    'public/javascripts/spacedeck_routes.js',
                    'public/javascripts/spacedeck_formatting.js',
                    'public/javascripts/spacedeck_sections.js',
                    'public/javascripts/spacedeck_spaces.js',
                    'public/javascripts/spacedeck_teams.js',
                    'public/javascripts/spacedeck_board_artifacts.js',
                    'public/javascripts/spacedeck_users.js',
                    'public/javascripts/spacedeck_account.js',
                    'public/javascripts/spacedeck_modals.js',
                    'public/javascripts/spacedeck_avatars.js',
                    'public/javascripts/spacedeck_websockets.js',

                    'public/javascripts/spacedeck_whiteboard.js',
                    'public/javascripts/spacedeck_directives.js',
                    'public/javascripts/spacedeck_vue.js',

                ],
                "stylesheets/style.css": [
                    'public/stylesheets/*.css'
                ]
            }
        }),
        new CopyWebpackPlugin([
            {
                from: 'public/{images,fonts}/**/*', to: 'build/',
                transformPath(targetPath, absolutePath) {
                    console.log('ta', targetPath);
                    return targetPath.replace(/build\/public\//, '');
                }
            },

            // { from: '**/*', to: '/absolute/path/to/dest/' }
        ], {})

    ],
    output: {
        filename: '[name].js',
        path: __dirname + '/build'
    }

};

middlewares/space_helpers.js
'use strict';

const db = require('../models/db');
var config = require('config');
const createError = require('http-errors')

/**
 * Not sure for the case:
 *  - user has a role in the room
 *  and uses edithash...?
 * @param space
 * @param user
 * @param userEditHash
 * @returns {Promise<*>}
 */
async function authenticateSpace(space, user, userEditHash) {

    if (userEditHash && space.edit_hash && userEditHash === space.edit_hash) {
        // Todo I think i should the edit hash here?
        return { anonyoums: true, role: 'editor' };
    }

    // handle public spaces
    if (space.access_mode === "public") {
        if (space.password && !req.spacePassword) {
            throw new createError(403, "password missing");
        }
        if (space.password && space.password !== req.spacePassword) {
            // todo: password not hashed!
            throw new createError(403, "password wrong");
        }
        return { anonyoums: true, role: 'viewer' };
    }

    // handle user is registered (and has a role in space)
    if (user) {
        const roleInSpace = await db.getUserRoleInSpacePromise(space, req.user);
        if (roleInSpace !== "none") {
            // const chosenLevel = getHigherRole(roleInSpace, userRequestAuth);
            return { anonyoums: false, role: roleInSpace };
        }
    }

    throw new createError(401, "not authenticated");


}

/**
 * Known roles:
 *  admin;editor;viewer
 * @param roleA
 * @param roleB
 * @deprecated
 * @returns {*}
 */
function getHigherRole(roleA, roleB) {
    if (roleA === "admin" || roleB === 'admin') {
        return "admin";
    }
    if (roleA === "editor" || roleB === 'editor') {
        return "editor";
    }
    return roleA;
}

/**
 * Comment: this module should be completely rewritten;
 * it's really tough to follow the logic here.
 *
 * @param req
 * @param res
 * @param next
 */
module.exports = async (req, res, next) => {
    let spaceId = req.params.id;
    const userRequestAuth = req.spaceAuth;
    console.log('userRequestAuth',req.spaceAuth, req.headers['x-spacedeck-space-auth'])

    const space = await db.Space.findOne({ where: { "_id": spaceId } })
    if (!space) {
        res.status(404).json({ "error": "space_not_found" });
        return;
    }

    // special permission for screenshot/pdf export from backend
    if (req.query['api_token'] && req.query['api_token'] == config.get('phantom_api_secret')) {
        finalizeReq(space, "viewer");
        return;
    }
    Promise.resolve()
        .then(() => authenticateSpace(space, req.user, userRequestAuth))
        .then(({ anonyoums, role }) => {
            if (role === "none") {
                throw new createError(403, "access denied");
            }
            req['space'] = space;
            req['spaceRole'] = role;
            console.log('setting role', role);
            res.header("x-spacedeck-space-role", req['spaceRole']);
            next();
        })
        .catch((error) => {
            res.send(error)
        })
    // });


}

routes/api/space_autocreate.js
const db = require('../../models/db');
// const uuidv4 = require('uuid/v4');
const express = require('express');
const router = express.Router({ mergeParams: true });
const crypto = require('crypto');

/*
  If the space does not exist;
  create it; make it public;
  redirect the user;
 */
router.get('/', async (req, res, next) => {


    const id = req.params.id;
    console.log('finidng ', id);
    let space = await db.Space.findOne({
        where: {
            "_id": id
        }
    });

    if (!space) {
        space = await createPublicSpace(id);
    }

    if (space.access_mode !== "public") {
        res.status(403).json({ "error": "no authorized" });
        return;
    }

    res.redirect("/spaces/" + id + "?spaceAuth=" + space.edit_hash);
});

async function createPublicSpace(id) {
    attrs = {};
    attrs._id = id;
    attrs.creator_id = null; // req.user._id;
    attrs.edit_hash = crypto.randomBytes(64).toString('hex').substring(0, 7);
    attrs.name = id;
    attrs.edit_slug = id;
    attrs.access_mode = 'public';
    let newSpace = await db.Space.create(attrs);

    return newSpace;


}

module.exports = router;


spacedeck.js
"use strict";

const db = require('./models/db.js');
require("log-timestamp");

const config = require('config');
const redis = require('./helpers/redis');
const websockets = require('./helpers/websockets');

const http = require('http');
const path = require('path');

const _ = require('underscore');
const favicon = require('serve-favicon');
const logger = require('morgan');
const cookieParser = require('cookie-parser');
const bodyParser = require('body-parser');

const swig = require('swig');
const i18n = require('i18n-2');
const helmet = require('helmet');

const express = require('express');
const app = express();
const serveStatic = require('serve-static');

const isProduction = app.get('env') === 'production';

console.log("Booting Spacedeck Open… (environment: " + app.get('env') + ")");

app.use(logger(isProduction ? 'combined' : 'dev'));

i18n.expressBind(app, {
  locales: ["en", "de", "fr"],
  defaultLocale: "en",
  cookieName: "spacedeck_locale",
  devMode: (app.get('env') == 'development')
});

swig.setDefaults({
  varControls: ["[[", "]]"] // otherwise it's not compatible with vue.js
});

swig.setFilter('cdn', function(input, idx) {
  return input;
});

app.engine('html', swig.renderFile);
app.set('view engine', 'html');

// @todo: unclear at the moment how the build/{views,assets} folders are created;
if (isProduction) {
//   app.set('views', path.join(__dirname, 'build', 'views'));
//   app.use(favicon(path.join(__dirname, 'build', 'assets', 'images', 'favicon.png')));
//   app.use(express.static(path.join(__dirname, 'build', 'assets')));s
} else {
  app.use(favicon(path.join(__dirname, 'public', 'images', 'favicon.png')));
  app.use(express.static(path.join(__dirname, 'public')));
}

app.set('views', path.join(__dirname, 'views'));

app.use(bodyParser.json({
  limit: '50mb'
}));

app.use(bodyParser.urlencoded({
  extended: false,
  limit: '50mb'
}));

app.use(cookieParser());
app.use(helmet.frameguard())
app.use(helmet.xssFilter())
app.use(helmet.hsts({
  maxAge: 7776000000,
  includeSubdomains: true
}))
app.disable('x-powered-by');
app.use(helmet.noSniff())

//app.use(require("./middlewares/error_helpers"));
app.use(require("./middlewares/session"));
//app.use(require("./middlewares/cors"));
app.use(require("./middlewares/i18n"));
app.use("/api", require("./middlewares/api_helpers"));
app.use('/api/spaces/:id', require("./middlewares/space_helpers"));
app.use('/api/spaces/:id/artifacts/:artifact_id', require("./middlewares/artifact_helpers"));

app.use('/api/users', require('./routes/api/users'));
app.use('/api/memberships', require('./routes/api/memberships'));

const spaceRouter = require('./routes/api/spaces');
app.use('/api/spaces', spaceRouter);

spaceRouter.use('/:id/artifacts', require('./routes/api/space_artifacts'));
spaceRouter.use('/:id/memberships', require('./routes/api/space_memberships'));
spaceRouter.use('/:id/messages', require('./routes/api/space_messages'));
app.use('/autocreate/:id', require('./routes/api/space_autocreate'));
spaceRouter.use('/:id/digest', require('./routes/api/space_digest'));
spaceRouter.use('/:id', require('./routes/api/space_exports'));

app.use('/api/sessions', require('./routes/api/sessions'));
//app.use('/api/webgrabber', require('./routes/api/webgrabber'));
app.use('/', require('./routes/root'));

if (config.get('storage_local_path')) {
  app.use('/storage', serveStatic(config.get('storage_local_path')+"/"+config.get('storage_bucket'), {
    maxAge: 24*3600
  }));
}

// catch 404 and forward to error handler
//app.use(require('./middlewares/404'));
if (app.get('env') === 'development') {
  app.set('view cache', false);
  swig.setDefaults({cache: false});
} else {
  app.use(require('./middlewares/500'));
}

module.exports = app;

// CONNECT TO DATABASE
db.init();

// START WEBSERVER
const port = 9666;

const server = http.Server(app).listen(port, () => {
  
  if ("send" in process) {
    process.send('online');
  }

}).on('listening', () => {
  
  const host = server.address().address;
  const port = server.address().port;
  console.log('Spacedeck Open listening at http://%s:%s', host, port);

}).on('error', (error) => {

  if (error.syscall !== 'listen') {
    throw error;
  }

  const bind = typeof port === 'string' ? 'Pipe ' + port : 'Port ' + port;
  switch (error.code) {
    case 'EACCES':
      console.error(bind + ' requires elevated privileges');
      process.exit(1);
      break;
    case 'EADDRINUSE':
      console.error(bind + ' is already in use');
      process.exit(1);
      break;
    default:
      throw error;
  }
});

websockets.startWebsockets(server);
redis.connectRedis();

/*process.on('message', (message) => {
  console.log("Process message:", message);
  if (message === 'shutdown') {
    console.log("Exiting Spacedeck.");
    process.exit(0);
  }
});*/
